armi/physics/fuelCycle/hexAssemblyFuelMgmtUtils.py

```diff
@@ -19,108 +19,113 @@
-        This orientation is relative to the current assembly orientation.
-        rot = 0 corresponds to no rotation.
-        rot represents the number of pi/3 counterclockwise rotations for the default orientation.
+    int
+        An integer from 0 to 5 representing the number of pi/3 (60 degree) counterclockwise
+        rotations from where ``a`` is currently oriented to the "optimal" orientation
 
-    Examples
-    --------
-    >>> getOptimalAssemblyOrientation(a, aPrev)
-    4
+    Raises
+    ------
+    ValueError
+        If there is insufficient information to determine the rotation of ``a``. This could
+        be due to a lack of fuel blocks or parameters like ``linPowByPin``.
 
-    See Also
-    --------
-    rotateAssemblies : calls this to figure out how to rotate
+    Notes
+    -----
+    Works by placing the highest-burnup pin in the location (of 6 possible locations) with lowest
+    expected pin power. We evaluated "expected pin power" based on the power distribution in
+    ``aPrev``, the previous assembly located where ``a`` is going. The algorithm goes as follows.
+
+    1. Get all the pin powers and ``IndexLocation``s from the block at the previous location.
+    2. Obtain the ``IndexLocation`` of the pin with the highest burnup in the current assembly.
+    3. For each possible rotation,
+        - Find the new location with ``HexGrid.rotateIndex``
+        - Find the index where that location occurs in previous locations
+        - Find the previous power at that location
+    4. Return the rotation with the lowest previous power
+
+    This algorithm assumes a few things.
+
+    1. ``len(HexBlock.getPinCoordinates()) == len(HexBlock.p.linPowByPin)`` and,
+       by extension, ``linPowByPin[i]`` is found at ``getPinCoordinates()[i]``.
+    2. Your assembly has at least 60 degree symmetry of fuel pins and
+       powers. This means if we find a fuel pin and rotate it 60 degrees, there should
+       be another fuel pin at that lattice site. This is mostly a safe assumption
+       since many hexagonal reactors have at least 60 degree symmetry of fuel pin layout.
+       This assumption holds if you have a full hexagonal lattice of fuel pins as well.
+    3. Fuel pins in ``a`` have similar locations in ``aPrev``. This is mostly a safe
+       assumption in that most fuel assemblies have similar layouts so it's plausible
+       that if ``a`` has a fuel pin at ``(1, 0, 0)``, so does ``aPrev``.
     """
-    # determine whether or not aPrev had pin details
-    fuelPrev = aPrev.getFirstBlock(Flags.FUEL)
-    if fuelPrev:
-        aPrevDetailFlag = fuelPrev.p.pinLocation[4] is not None
+    maxBuBlock = maxBurnupBlock(a)
+    if maxBuBlock.spatialGrid is None:
+        msg = f"Block {maxBuBlock} in {a} does not have a spatial grid. Cannot rotate."
+        runLog.error(msg)
+        raise ValueError(msg)
+    maxBuPinLocation = maxBurnupLocator(maxBuBlock)
```

Comment:
```suggestion
        raise ValueError(msg)

    # 2. Obtain the ``IndexLocation`` of the pin with the highest burnup in the current assembly.
    maxBuPinLocation = maxBurnupLocator(maxBuBlock)
```

Reply:
If it's my call, I'm going to forgo this and similar requests.

I get readability through comments. I've also seen comments rot as the code changes while comments are not updated accordingly. So I'm inclined to have more verbose variable names and functions over comments. To hopefully help explain things.

`maxBuPinLocation` _feels_ to me like a thing that tells us where the pin with the highest burnup can be found

It's very much a personal stance of mine that, if I'm left to decide, I'm going to lean on fewer comments but more descriptive docstrings and verbose variables/functions

Reply:
Okay, that's fine.

That's not the M.O. for ARMI though. I don't particularly feel that docstrings are less likely to go out of date than comments.  And without your docstring, this function's logic would be too complicated for someone to read at a glance. That's really my problem, I don't want some poor engineer to come to this function in 3 years and have to puzzle out this important logic on their own.

To that end, the docstring can stand as sufficient.